hi-ml/src/health_ml/networks/layers/attention_layers.py

```diff
@@ -89,3 +86,161 @@ def forward(self, features: Tensor) -> Tuple[Tensor, Tensor]:
+    """
+    __constants__ = ['batch_first', 'norm_first']
+
+    def __init__(self, d_model: int,
+                 nhead: int,
+                 dim_feedforward: int = 2048,
+                 dropout: float = 0.1,
+                 activation: Callable = F.relu,
+                 layer_norm_eps: float = 1e-5,
+                 batch_first: bool = True,
+                 norm_first: bool = False,
+                 device: Optional[torch.device] = None,
+                 dtype: Optional[torch.dtype] = None) -> None:
+        factory_kwargs = {'device': device, 'dtype': dtype}
+        super(CustomTransformerEncoderLayer, self).__init__()
+        self.self_attn = MultiheadAttention(d_model, nhead, dropout=dropout, batch_first=batch_first,
+                                            **factory_kwargs)
+        # Implementation of Feedforward model
+        self.linear1 = Linear(d_model, dim_feedforward, **factory_kwargs)  # type: ignore
+        self.dropout = Dropout(dropout)
+        self.linear2 = Linear(dim_feedforward, d_model, **factory_kwargs)  # type: ignore
+
+        self.norm_first = norm_first
+        self.norm1 = LayerNorm(d_model, eps=layer_norm_eps, **factory_kwargs)  # type: ignore
+        self.norm2 = LayerNorm(d_model, eps=layer_norm_eps, **factory_kwargs)  # type: ignore
+        self.dropout1 = Dropout(dropout)
+        self.dropout2 = Dropout(dropout)
+        self.activation = activation
+
+    def forward(self, src: torch.Tensor,
+                src_mask: Optional[torch.Tensor] = None,
+                src_key_padding_mask: Optional[torch.Tensor] = None) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Pass the input through the encoder layer.
+
+        Args:
+            src: the sequence to the encoder layer (required).
+            src_mask: the mask for the src sequence (optional).
+            src_key_padding_mask: the mask for the src keys per batch (optional).
+
+        Shape:
+            see the docs in Transformer class.
+        """
+
+        # see Fig. 1 of https://arxiv.org/pdf/2002.04745v1.pdf
+
+        x = src
+        if self.norm_first:
+            sa_block_out, a = self._sa_block(self.norm1(x), src_mask, src_key_padding_mask)
+            x = x + sa_block_out
+            x = x + self._ff_block(self.norm2(x))
+        else:
+            sa_block_out, a = self._sa_block(x, src_mask, src_key_padding_mask)
+            x = self.norm1(x + sa_block_out)
+            x = self.norm2(x + self._ff_block(x))
+
+        return x, a
+
+    # self-attention block
+    def _sa_block(self, x: Tensor,
+                  attn_mask: Optional[Tensor], key_padding_mask: Optional[Tensor]) -> Tuple[torch.Tensor, torch.Tensor]:
+        x, a = self.self_attn(x, x, x,
+                              attn_mask=attn_mask,
+                              key_padding_mask=key_padding_mask,
+                              need_weights=True)  # Just because of this flag I had to copy all of the code...
```

Comment:
Is there no cleaner way to apply this change than replicating the entire class? How about subclassing `TransformerEncoderLayer` and overriding just this `_sa_block()` method?

Reply:
Much better! Minor optional suggestion to add a comment explaining that we needed to also override `forward()` to return the attentions.